### ISSUE TITLE
False positive UndefinedThisPropertyFetch when using $this->magicProperty in method

### ISSUE THREAD

**Issue by TysonAndre** [vimeo/psalm#798]
```php
<?php
/**
 * @property-read string $name
 * @property string $otherName
 */
class A {
  public function __get(string $name): void {
  }

  public function goodGet(): void {
    echo $this->name;
  }
  public function goodGet2(): void {
    echo $this->otherName;
  }
}
$a = new A();
echo $a->name;
echo $a->otherName;
```

Expected: no error
Observed:
```
ERROR: UndefinedThisPropertyFetch - src/x.php:11:10 - Instance property A::$name is not defined
    echo $this->name;


ERROR: MixedArgument - src/x.php:11:10 - Argument 1 of echo cannot be mixed, expecting string
    echo $this->name;


ERROR: UndefinedThisPropertyFetch - src/x.php:14:10 - Instance property A::$otherName is not defined
    echo $this->otherName;


ERROR: MixedArgument - src/x.php:14:10 - Argument 1 of echo cannot be mixed, expecting string
    echo $this->otherName;
```

See https://github.com/vimeo/psalm/issues/499#issuecomment-364112040

c1440c11dcc559d65aa0f93921240a5f8a6cf538 introduced directFetchForMagicProperty, which I think should instead **not** throw. The misnamedPropertyByVariable test is a bug, but would require more complicated logic of checking if property name in __get() didn't change at all.
